pkg/cmd/server/kubernetes/master.go

```diff
@@ -131,14 +138,52 @@ func (c *MasterConfig) InstallAPI(container *restful.Container) ([]string, error
 // RunNamespaceController starts the Kubernetes Namespace Manager
 func (c *MasterConfig) RunNamespaceController(kubeClient internalclientset.Interface, clientPool dynamic.ClientPool) {
 	// Find the list of namespaced resources via discovery that the namespace controller must manage
-	groupVersionResources, err := namespacecontroller.ServerPreferredNamespacedGroupVersionResources(kubeClient.Discovery())
+	groupVersionResources, err := kubeClient.Discovery().ServerPreferredNamespacedResources()
 	if err != nil {
 		glog.Fatalf("Failed to get supported resources from server: %v", err)
 	}
 	namespaceController := namespacecontroller.NewNamespaceController(kubeClient, clientPool, groupVersionResources, c.ControllerManager.NamespaceSyncPeriod.Duration, kapi.FinalizerKubernetes)
 	go namespaceController.Run(int(c.ControllerManager.ConcurrentNamespaceSyncs), utilwait.NeverStop)
 }
 
+func (c *MasterConfig) RunPersistentVolumeController(client *client.Client, namespace, recyclerImageName string) {
+	s := c.ControllerManager
+	provisioner, err := kctrlmgr.NewVolumeProvisioner(c.CloudProvider, s.VolumeConfiguration)
+	if err != nil {
+		glog.Fatal("A Provisioner could not be created, but one was expected. Provisioning will not work. This functionality is considered an early Alpha version.")
+	}
+
+	volumeController := persistentvolumecontroller.NewPersistentVolumeController(
+		clientadapter.FromUnversionedClient(client),
+		s.PVClaimBinderSyncPeriod.Duration,
+		provisioner,
+		probeRecyclableVolumePlugins(s.VolumeConfiguration, namespace, recyclerImageName),
+		c.CloudProvider,
+		s.ClusterName,
+		nil, nil, nil,
+		s.VolumeConfiguration.EnableDynamicProvisioning,
+	)
+	volumeController.Run()
+	time.Sleep(utilwait.Jitter(s.ControllerStartInterval.Duration, kctrlmgr.ControllerStartJitter))
+
+	attachDetachController, err :=
+		volumecontroller.NewAttachDetachController(
+			clientadapter.FromUnversionedClient(client),
+			c.Informers.Pods().Informer(),
+			c.Informers.Nodes().Informer(),
+			c.Informers.PersistentVolumeClaims().Informer(),
+			c.Informers.PersistentVolumes().Informer(),
+			c.CloudProvider,
+			kctrlmgr.ProbeAttachableVolumePlugins(s.VolumeConfiguration))
+	if err != nil {
+		glog.Fatalf("Failed to start attach/detach controller: %v", err)
+	} else {
+		go attachDetachController.Run(utilwait.NeverStop)
+		time.Sleep(utilwait.Jitter(s.ControllerStartInterval.Duration, kctrlmgr.ControllerStartJitter))
```

Comment:
"Doesn't matter" in the sense that "this won't actually do anything, so the
code here doesn't matter".  The controller won't do any work until the
shared informer starts its watch and that doesn't happen until all
controllers are initialized, so all this code accomplishes is slowing down
server start.  You still get a storm.

On Fri, Jul 8, 2016 at 5:31 PM, Clayton Coleman <notifications@github.com>
wrote:

> In pkg/cmd/server/kubernetes/master.go
> <https://github.com/openshift/origin/pull/9578#discussion_r70144055>:
>
> > +	time.Sleep(utilwait.Jitter(s.ControllerStartInterval.Duration, kctrlmgr.ControllerStartJitter))
> > +
> > +	attachDetachController, err :=
> > +		volumecontroller.NewAttachDetachController(
> > +			clientadapter.FromUnversionedClient(client),
> > +			c.Informers.Pods().Informer(),
> > +			c.Informers.Nodes().Informer(),
> > +			c.Informers.PersistentVolumeClaims().Informer(),
> > +			c.Informers.PersistentVolumes().Informer(),
> > +			c.CloudProvider,
> > +			kctrlmgr.ProbeAttachableVolumePlugins(s.VolumeConfiguration))
> > +	if err != nil {
> > +		glog.Fatalf("Failed to start attach/detach controller: %v", err)
> > +	} else {
> > +		go attachDetachController.Run(utilwait.NeverStop)
> > +		time.Sleep(utilwait.Jitter(s.ControllerStartInterval.Duration, kctrlmgr.ControllerStartJitter))
>
> It actually does matter, even with a shared informer, because the end
> result of that could be a flurry of operations from the core controller.
> Jitter wasn't just about the core list/watch, but also about a sudden
> operation storm layering on top when every controller in the system boots
> all at once. I'll remove for now but we'll need to come back to this.
> … <#m_-3718658231585956686_>
> On Fri, Jul 8, 2016 at 11:31 AM, David Eads ***@***.***> wrote: In
> pkg/cmd/server/kubernetes/master.go <#9578 (comment)
> <https://github.com/openshift/origin/pull/9578#discussion_r70093166>>: >
> + time.Sleep(utilwait.Jitter(s.ControllerStartInterval.Duration,
> kctrlmgr.ControllerStartJitter)) > + > + attachDetachController, err := > +
> volumecontroller.NewAttachDetachController( > +
> clientadapter.FromUnversionedClient(client), > +
> c.Informers.Pods().Informer(), > + c.Informers.Nodes().Informer(), > +
> c.Informers.PersistentVolumeClaims().Informer(), > +
> c.Informers.PersistentVolumes().Informer(), > + c.CloudProvider, > +
> kctrlmgr.ProbeAttachableVolumePlugins(s.VolumeConfiguration)) > + if err !=
> nil { > + glog.Fatalf("Failed to start attach/detach controller: %v", err)
> > + } else { > + go attachDetachController.Run(utilwait.NeverStop) > +
> time.Sleep(utilwait.Jitter(s.ControllerStartInterval.Duration,
> kctrlmgr.ControllerStartJitter)) we haven't done this elsewhere, I'm not
> crazy about the idea and it doesn't work with a SharedInformer. — You are
> receiving this because you authored the thread. Reply to this email
> directly, view it on GitHub <
> https://github.com/openshift/origin/pull/9578/files/3c648ae1593e9190dd435d5d7fcd0f249fee5afa..850623848185fb0a2060162f16a198dcbfe0d1c1#r70093166>,
> or mute the thread <
> https://github.com/notifications/unsubscribe/ABG_p0y3BJvZDLOdYYQU-4US3vUnIkk4ks5qTm1lgaJpZM4I-xTE>
> .
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/9578/files/3c648ae1593e9190dd435d5d7fcd0f249fee5afa..850623848185fb0a2060162f16a198dcbfe0d1c1#r70144055>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AH2BSlg7FNUUBJPUWK0aiWMD2JFNL_d_ks5qTsGfgaJpZM4I-xTE>
> .
>